File: src/main/java/AdaBoost/AdaBoost.java
package AdaBoost;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import util.ValUtil;

public class AdaBoost {
	
	public List<String> inequalList = Arrays.asList("lt","gt");
	
	public double[] stumpClassify(double[][] dataMatrix,int dimen,double threshVal,String threshIneq) {
		double [] retArray = new double[dataMatrix.length];
		for(int i=0;i<retArray.length;i++) {
			retArray[i] = 1;
		}
		
		if("lt".equals(threshIneq)) {
			for(int i=0;i<retArray.length;i++) {
				if(dataMatrix[i][dimen] <= threshVal) {
					retArray[i] = -1;
				}
			}
		}else {
			for(int i=0;i<retArray.length;i++) {
				if(dataMatrix[i][dimen] > threshVal) {
					retArray[i] = -1;
				}
			}
		}
		return retArray;
	}
	
	public Map<String,Object> buildStump(double [][] dataArr,double [] classLabels, double [] D) {
		
		Map<String,Object> retMap = new HashMap<String, Object>();
		
		double [][] dataMatrix = dataArr;
		double [] labelMat = classLabels;
		int m = dataMatrix.length;
		int n = dataMatrix[0].length;
		
		double numSteps = 10.0;
		Map<String,Object> bestStump = new HashMap<String,Object>();
		double [] bestClasEst = new double[m];
		
		double minError = Double.MAX_VALUE;
		
		for(int i=0;i<n;i++) {
			double rangeMin = Double.MAX_VALUE;
			double rangeMax = -Double.MAX_VALUE;// min
			for(int j=0;j<m;j++) {
				if(rangeMin > dataMatrix[j][i]) {
					rangeMin = dataMatrix[j][i];
				}
				if(rangeMax < dataMatrix[j][i]) {
					rangeMax = dataMatrix[j][i];
				}
			}
			
			double stepSize = (rangeMax - rangeMin)/numSteps;
			
			for(int j=-1;j<((int)numSteps + 1);j++) {
				
				for(String inequal:inequalList) {
					double threshVal = (rangeMin + (j*1.0)*stepSize);
					
					double[] predictedVals = stumpClassify(dataMatrix, i, threshVal, inequal);
					
					int [] errArr = new int[m];
					for(int k = 0; k < m; k++) {
						if(predictedVals[k] == labelMat[k]) {
							errArr[k] = 0;
						}else {
							errArr[k] = 1;
						}
					}
					
					double weightedError = 0;
					for(int k = 0; k < m; k++) {
						weightedError += D[k] * errArr[k];
					}
					
					if(weightedError < minError) {
						minError = weightedError;
						bestClasEst = predictedVals.clone();
						bestStump.put("dim", i);
						bestStump.put("thresh", threshVal);
						bestStump.put("ineq", inequal);
					}
				}
			}
		}
		
		retMap.put("bestStump", bestStump);
		retMap.put("minError", minError);
		retMap.put("bestClasEst", bestClasEst);
		
		return retMap;
	}
	
	@SuppressWarnings("unchecked")
	public List<Map<String,Object>> adaBoostTrainDS(double [][] dataArr,double [] classLabels, int numIt) {
		
		List<Map<String,Object>> weakClassArr = new ArrayList<Map<String,Object>>();
		int m = dataArr.length;
		double [] D = new double[m];
		for(int i=0;i<m;i++) {
			D[i] = 1.0/m;
		}
		double [] aggClassEst = new double[m];
		
		for(int i=1;i<numIt;i++) {
			Map<String,Object> stumpObject = buildStump(dataArr, classLabels, D);
			Map<String,Object> bestStump = (Map<String, Object>) stumpObject.get("bestStump");
			double error = (Double) stumpObject.get("minError");
			double [] classEst = (double[]) stumpObject.get("bestClasEst");
			
			double alpha = 0.5*Math.log((1.0 - error)/Math.max(error, Double.MIN_VALUE));
			bestStump.put("alpha", alpha);
			
			weakClassArr.add(bestStump);
			
			double[] expon = new double[m];
			double dSum = 0;
			for(int j=0;j<m;j++) {
				expon[j] = -1 * alpha * classLabels[j] * classEst[j];
				D[j] = D[j] * (Math.pow(Math.E, expon[j]));
				dSum += D[j];
			}
			
			for(int j=0;j<m;j++) {
				D[j] = (1.0*D[j])/dSum;
			}
			
			for(int j=0;j<m;j++) {
				aggClassEst[j] += alpha * classEst[j];
			}
			
			int [] aggErrors = new int[m];
			int aggErrorsSum = 0;
			for(int j=0;j<m;j++) {
				if(ValUtil.sign(aggClassEst[j]) != classLabels[j]) {
					aggErrors[j] = 1;
				}else {
					aggErrors[j] = 0;
				}
				aggErrorsSum += aggErrors[j]; 
			}
			double errorRate = (1.0*aggErrorsSum)/m;
			System.out.println("total error："+errorRate);
			
			if(errorRate == 0.0)
				break;
		}
		
		return weakClassArr;
	}
	
	public double[] adaClassify(double [][] datToClass, List<Map<String,Object>> classifierArr) {
		double [][] dataMatrix = datToClass;
		int m = dataMatrix.length;
		
		double [] aggClassEst = new double[m];
		
		for (int i=0;i<classifierArr.size();i++) {
			double[] classEst = stumpClassify(dataMatrix,
					(Integer)classifierArr.get(i).get("dim"), 
					(Double)classifierArr.get(i).get("thresh"), 
					(String)classifierArr.get(i).get("ineq"));
			
			for(int j=0;j<m;j++) {
				aggClassEst[j] += ((Double)classifierArr.get(i).get("alpha")) * classEst[j];
			}
		}
		
		double[] ret = new double[m];
		for(int j=0;j<m;j++) {
			ret[j] = ValUtil.sign(aggClassEst[j]);
		}
		return ret;
	}
}

File: src/main/java/AdaBoost/adaboost.py
#-*- coding: utf-8 -*-

'''
Created on Nov 28, 2010
Adaboost is short for Adaptive Boosting
@author: Peter
'''
from numpy import *

# 读取训练数据到数据矩阵和类别标签列表
def loadDataSet(filename):
    dataMat = []; labelMat = []
    
    # 打开文件
    fr = open(filename)
    
    # 遍历数据文件每一行
    for line in fr.readlines():
        
        # 根据制表符切分每一行的数据
        currLine = line.strip().split('\t')
        lineArr =[]
        
        # 将22项数据都去进lineArr列表
        for i in range(21):
            lineArr.append(float(currLine[i]))
        
        # 添加到矩阵和类别标签列表
        dataMat.append(lineArr)
        
        # 第22列为类别
        labelMat.append(float(currLine[21]))
    return dataMat, labelMat
    
# 分类函数，通过输入的阈值参数进行分类
def stumpClassify(dataMatrix,dimen,threshVal,threshIneq):

    # 初始化分类结果为1
    retArray = ones((shape(dataMatrix)[0],1))
    
    # 判断阈值中条件
    if threshIneq == 'lt':
        # 条件为小于等于阈值的类别为-1
        retArray[dataMatrix[:,dimen] <= threshVal] = -1.0
    else:
        retArray[dataMatrix[:,dimen] > threshVal] = -1.0
    return retArray

# 单层决策树生成函数，找到并建立数据集上的最佳的单层决策树
def buildStump(dataArr,classLabels,D):
    # 构建输入数据矩阵和类别标签矩阵
    dataMatrix = mat(dataArr); labelMat = mat(classLabels).T
    m,n = shape(dataMatrix)
    
    # 创建bestStump用于存储最佳单层决策树的相关信息
    numSteps = 10.0; bestStump = {}; bestClasEst = mat(zeros((m,1)))
    
    # 初始化错误率为无穷大
    minError = inf
    
    # 循环遍历数据集的所有特征
    for i in range(n):
        # 通过计算数据最大最小值来获取步长
        rangeMin = dataMatrix[:,i].min(); rangeMax = dataMatrix[:,i].max();
        stepSize = (rangeMax-rangeMin)/numSteps
        
        # 在当前特征对应的数据上进行遍历
        for j in range(-1,int(numSteps)+1):
        
            # 再大于和小于之间切换不等式条件
            for inequal in ['lt', 'gt']:

                # 得到本次执行的阈值
                threshVal = (rangeMin + float(j) * stepSize)
                
                # 获得数据分类结果
                predictedVals = stumpClassify(dataMatrix,i,threshVal,inequal)
                
                # 计算错误率
                errArr = mat(ones((m,1)))
                errArr[predictedVals == labelMat] = 0
                weightedError = D.T*errArr
                
                # 判断加权错误率是否小于当前的最小错误率
                if weightedError < minError:
                    # 更新决策树
                    minError = weightedError
                    bestClasEst = predictedVals.copy()
                    bestStump['dim'] = i
                    bestStump['thresh'] = threshVal
                    bestStump['ineq'] = inequal
    # 返回最佳单层决策树
    return bestStump,minError,bestClasEst

# AdaBoost训练过程，返回训练结果为弱分类器
def adaBoostTrainDS(dataArr,classLabels,numIt=40):
    
    # 初始化要返回的弱分类器列表
    weakClassArr = []
    m = shape(dataArr)[0]
    
    # 初始化D值，赋予相等的权重
    D = mat(ones((m,1))/m)
    aggClassEst = mat(zeros((m,1)))
    
    # 开始训练迭代
    for i in range(numIt):
    
        # 获取最佳单层决策树，最小错误率和估计的类别向量
        bestStump,error,classEst = buildStump(dataArr,classLabels,D)
        
        # 计算alpha值，为本次分类器输出结果的权重
        alpha = float(0.5*log((1.0-error)/max(error,1e-16)))
        bestStump['alpha'] = alpha
        
        # 将本次获得的决策树存入分类器列表
        weakClassArr.append(bestStump)
        
        # 计算下一次迭代的D值
        expon = multiply(-1*alpha*mat(classLabels).T,classEst) 
        D = multiply(D,exp(expon))
        D = D/D.sum()
        
        # 计算总分类器的错误率，aggClassEst保存运行时的类型估计值
        aggClassEst += alpha*classEst
        aggErrors = multiply(sign(aggClassEst) != mat(classLabels).T,ones((m,1)))
        errorRate = aggErrors.sum()/m
        print "total error: ",errorRate
        
        # 如果错误率已经为0则退出迭代
        if errorRate == 0.0: break
    
    # 返回弱分类器
    return weakClassArr

# AdaBoost分类函数，根据训练得到的弱分类器对数据进行分类
def adaClassify(datToClass,classifierArr):
    # 初始化输入数据
    dataMatrix = mat(datToClass)
    m = shape(dataMatrix)[0]
    
    # 初始化aggClassEst为全0向量
    aggClassEst = mat(zeros((m,1)))
    
    # 依次使用每个弱分类器
    for i in range(len(classifierArr)):
    
        # 使用弱分类器得到类别估计值
        classEst = stumpClassify(dataMatrix,classifierArr[i]['dim'],\
                                 classifierArr[i]['thresh'],\
                                 classifierArr[i]['ineq'])
        
        # 输出的类别乘以权重后累加到aggClassEst上
        aggClassEst += classifierArr[i]['alpha']*classEst
    
    # 返回sign()函数处理过的aggClassEst值即为预计的类别
    return sign(aggClassEst)
    
# 测试算法
def adaTest():
    # 读取训练数据
    trainingSet, trainingLabels = loadDataSet('horseColicTraining2.txt')
    
    # 训练算法
    classifierArray = adaBoostTrainDS(trainingSet, trainingLabels)

    # 测试条目数量
    numTestVec = 67.0
    
    # 读取测试数据
    testSet, testLabels = loadDataSet('horseColicTest2.txt')
    
    # 使用分类器进行预测
    prediction10 = adaClassify(testSet, classifierArray)
    
    # 对比分类预测结果及真实结果并记录错误率
    errArr = mat(ones((67,1)))
    
    # 矩阵计算得到判断出错的条目数
    errorCount = errArr[prediction10!=mat(testLabels).T].sum()
    
    # 计算并输出错误率
    errorRate = (float(errorCount)/numTestVec)
    print "the error rate of this test is: %f" % errorRate
    return errorRate
    
    
adaTest()
    
    